bin/main.rb

```diff
@@ -103,40 +113,63 @@ def display_board(cells)
 name1 = ui.prompt_player_name('player 1')
 symbol1 = 'X'
 array1 = []
-player1 = Player.new(name1, symbol1, array1)
+player1 = Player.new(name1.magenta, symbol1, array1)
 name2 = ui.prompt_player_name('player 2')
 symbol2 = 'O'
 array2 = []
-player2 = Player.new(name2, symbol2, array2)
-current_player = 'player1'
+player2 = Player.new(name2.cyan, symbol2, array2)
+again="Y"
 
-board = Board.new(current_player)
-game_logic = GameLogic.new(player1, player2, board.cell)
-ui.display_board board.cell
-
-until game_logic.game_end
+  while again == "Y"
+  current_player = 'player1'
+  board = Board.new(current_player)
+  game_logic = GameLogic.new(player1, player2, board.cell)
+  ui.display_board board.cell
+  moves = 0 
+  until game_logic.game_end || moves >8
+
+    mark = ui.get_mark(if current_player == 'player1'
+                        player1.name
+                      else
+                        player2.name
+                      end).to_i
+      
+    while mark==0 || mark > 9 || board.available?(mark)
+      if mark==0 || mark > 9
+        mark = ui.invalid_cell
+      else
+        mark = ui.display_available
+      end
+    end
 
-  mark = ui.get_mark(if current_player == 'player1'
-                       player1.name
-                     else
-                       player2.name
-                     end).to_i
 
-  if current_player == 'player1'
-    player1.array.push(mark)
-  else
-    player2.array.push(mark)
-  end
-  board.update_cell mark
-  board.change
+    if current_player == 'player1'
+      player1.array.push(mark)
+    else
+      player2.array.push(mark)
+    end
```

Comment:
This can be handled in the game logic,  why place it here? if you are to save the player moves, it should be handled in by the game

Reply:
These codes up there should be handled by the game class, you want to limit the logic in this file, the only thing needed here is the loop that runs the game, based on set conditions